File: capacitor-web/ios/App/App/MlKitTranslationPlugin.swift
import Capacitor
import MLKitCommon
import MLKitTranslate

@objc(MlKitTranslationPlugin)
public class MlKitTranslationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MlKitTranslationPlugin"
    public let jsName = "MlKitTranslation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "translate", returnType: CAPPluginReturnPromise)
    ]

    @objc func translate(_ call: CAPPluginCall) {
        guard let text = call.getString("text"), !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let sourceTag = call.getString("sourceLanguage"),
              let targetTag = call.getString("targetLanguage"),
              let source = TranslateLanguage.fromLanguageTag(sourceTag),
              let target = TranslateLanguage.fromLanguageTag(targetTag) else {
            call.reject("번역할 내용과 ML Kit 지원 언어를 확인해 주세요.")
            return
        }

        let options = TranslatorOptions(sourceLanguage: source, targetLanguage: target)
        let translator = Translator.translator(options: options)
        let conditions = ModelDownloadConditions(allowsCellularAccess: false, allowsBackgroundDownloading: true)
        translator.downloadModelIfNeeded(with: conditions) { error in
            if let error = error {
                call.reject("ML Kit 번역 모델을 내려받지 못했습니다.", nil, error)
                return
            }
            translator.translate(text) { translatedText, error in
                if let error = error {
                    call.reject("ML Kit 번역을 실행하지 못했습니다.", nil, error)
                    return
                }
                guard let translatedText = translatedText else {
                    call.reject("ML Kit 번역 결과가 비어 있습니다.")
                    return
                }
                call.resolve(["translation": translatedText])
            }
        }
    }
}

File: capacitor-web/android/app/src/main/java/org/dari/app/MlKitTranslationPlugin.java
package org.dari.app;

import androidx.annotation.NonNull;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.nl.translate.TranslateLanguage;
import com.google.mlkit.nl.translate.Translation;
import com.google.mlkit.nl.translate.Translator;
import com.google.mlkit.nl.translate.TranslatorOptions;

@CapacitorPlugin(name = "MlKitTranslation")
public class MlKitTranslationPlugin extends Plugin {
    @com.getcapacitor.PluginMethod
    public void translate(PluginCall call) {
        String text = call.getString("text", "").trim();
        String sourceTag = call.getString("sourceLanguage", "");
        String targetTag = call.getString("targetLanguage", "");
        String source = TranslateLanguage.fromLanguageTag(sourceTag);
        String target = TranslateLanguage.fromLanguageTag(targetTag);
        if (text.isEmpty()) {
            call.reject("번역할 내용이 없습니다.");
            return;
        }
        if (source == null || target == null) {
            call.reject("ML Kit이 지원하지 않는 언어입니다.");
            return;
        }

        Translator translator = Translation.getClient(new TranslatorOptions.Builder()
            .setSourceLanguage(source)
            .setTargetLanguage(target)
            .build());
        DownloadConditions conditions = new DownloadConditions.Builder().build();
        translator.downloadModelIfNeeded(conditions)
            .addOnSuccessListener(ignored -> translator.translate(text)
                .addOnSuccessListener(translated -> resolveAndClose(call, translator, translated))
                .addOnFailureListener(error -> rejectAndClose(call, translator, error)))
            .addOnFailureListener(error -> rejectAndClose(call, translator, error));
    }

    private void resolveAndClose(@NonNull PluginCall call, @NonNull Translator translator, String translated) {
        translator.close();
        JSObject result = new JSObject();
        result.put("translation", translated);
        call.resolve(result);
    }

    private void rejectAndClose(@NonNull PluginCall call, @NonNull Translator translator, Exception error) {
        translator.close();
        call.reject("ML Kit 번역 모델을 준비하거나 실행하지 못했습니다.", error);
    }
}
